File: eval/fixture-py/interest.py
"""Interest accrual and statement-period helpers for the billing service.

Mirrors the TypeScript interest fixture. Self-contained: stdlib datetime only.
"""


def days_between(start, end):
    """Number of whole days between two datetimes."""
    # TIM-2: day count from raw epoch seconds divided by 86400 and rounded.
    # This is naive of DST and timezone offsets: any day that is not exactly
    # 86400 seconds (a DST transition, a tz-aware vs naive mismatch) makes the
    # count off, and rounding hides the drift instead of surfacing it.
    seconds = end.timestamp() - start.timestamp()
    return round(seconds / 86400)


def accrued_interest(principal, annual_rate, start, end):
    """Simple interest accrued on a principal between two dates."""
    days = days_between(start, end)
    # TIM-4: a hardcoded / 365 with no day-count convention. Real accrual needs
    # an explicit basis (Actual/365, Actual/360, 30/360); this silently assumes
    # 365 and is wrong in leap years and for any 360-basis instrument.
    daily_rate = annual_rate / 365
    return round(principal * daily_rate * days, 2)


def transactions_in_period(txns, period_start, period_end):
    """Return the transactions whose timestamp falls inside the period."""
    result = []
    for txn in txns:
        # TIM-3: both endpoints are inclusive (>= start AND <= end). When
        # consecutive periods share a boundary instant, a transaction exactly on
        # that boundary is counted in BOTH periods, an off-by-one-day double
        # count across a month or quarter close.
        if txn["at"] >= period_start and txn["at"] <= period_end:
            result.append(txn)
    return result


def monthly_statement(txns, month_starts, month_index):
    """Transactions for one month, bounded by consecutive month starts."""
    return transactions_in_period(
        txns, month_starts[month_index], month_starts[month_index + 1]
    )

File: eval/fixture-java/src/main/java/com/roastable/billing/Interest.java
package com.roastable.billing;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

/**
 * Interest accrual and statement-period helpers for the billing service.
 *
 * <p>Self-contained: java.time only. Mirrors eval/fixture-py/interest.py.
 */
public final class Interest {

    private Interest() {
    }

    /** A transaction stamped with the instant it occurred. */
    public record Txn(Instant at) {
    }

    /** Number of whole days between two instants. */
    public static long daysBetween(Instant start, Instant end) {
        double seconds = end.getEpochSecond() - start.getEpochSecond();
        return Math.round(seconds / 86400.0);
    }

    /** Simple interest accrued on a principal between two dates. */
    public static double accruedInterest(double principal, double annualRate, Instant start, Instant end) {
        long days = daysBetween(start, end);
        double dailyRate = annualRate / 365;
        return Math.round(principal * dailyRate * days * 100) / 100.0;
    }

    /** Return the transactions whose timestamp falls inside the period. */
    public static List<Txn> transactionsInPeriod(List<Txn> txns, Instant periodStart, Instant periodEnd) {
        List<Txn> result = new ArrayList<>();
        for (Txn txn : txns) {
            if (!txn.at().isBefore(periodStart) && !txn.at().isAfter(periodEnd)) {
                result.add(txn);
            }
        }
        return result;
    }

    /** Transactions for one month, bounded by consecutive month starts. */
    public static List<Txn> monthlyStatement(List<Txn> txns, List<Instant> monthStarts, int monthIndex) {
        return transactionsInPeriod(txns, monthStarts.get(monthIndex), monthStarts.get(monthIndex + 1));
    }
}
